# Solution to Session #1 Homework

## Question 1
What's the version of NumPy that you installed?

In [72]:
import numpy as np
import pandas as pd

In [73]:
np.__version__

'1.19.4'

version of Pandas (for myself)

In [74]:
pd.__version__

'1.1.5'

### Getting the data
For this homework, we'll use the same dataset as for the next session - the car price dataset. 
Download it with: 

wget https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv



Then read it with Pandas


In [75]:
#!wget https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv

## Question 2
How many records are in the dataset?
Here we need to specify the number of rows.

**DataFrame**

In [76]:
df = pd.read_csv('data.csv')  #reading the data and creating the DataFrame

In [77]:
df

Make       Model  Year                Engine Fuel Type  Engine HP  \
0          BMW  1 Series M  2011     premium unleaded (required)      335.0   
1          BMW    1 Series  2011     premium unleaded (required)      300.0   
2          BMW    1 Series  2011     premium unleaded (required)      300.0   
3          BMW    1 Series  2011     premium unleaded (required)      230.0   
4          BMW    1 Series  2011     premium unleaded (required)      230.0   
...        ...         ...   ...                             ...        ...   
11909    Acura         ZDX  2012     premium unleaded (required)      300.0   
11910    Acura         ZDX  2012     premium unleaded (required)      300.0   
11911    Acura         ZDX  2012     premium unleaded (required)      300.0   
11912    Acura         ZDX  2013  premium unleaded (recommended)      300.0   
11913  Lincoln      Zephyr  2006                regular unleaded      221.0   

       Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
0                   6.0            MANUAL   rear wheel drive              2.0   
1                   6.0            MANUAL   rear wheel drive              2.0   
2                   6.0            MANUAL   rear wheel drive              2.0   
3                   6.0            MANUAL   rear wheel drive              2.0   
4                   6.0            MANUAL   rear wheel drive              2.0   
...                 ...               ...                ...              ...   
11909               6.0         AUTOMATIC    all wheel drive              4.0   
11910               6.0         AUTOMATIC    all wheel drive              4.0   
11911               6.0         AUTOMATIC    all wheel drive              4.0   
11912               6.0         AUTOMATIC    all wheel drive              4.0   
11913               6.0         AUTOMATIC  front wheel drive              4.0   

                             Market Category Vehicle Size  Vehicle Style  \
0      Factory Tuner,Luxury,High-Performance      Compact          Coupe   
1                         Luxury,Performance      Compact    Convertible   
2                    Luxury,High-Performance      Compact          Coupe   
3                         Luxury,Performance      Compact          Coupe   
4                                     Luxury      Compact    Convertible   
...                                      ...          ...            ...   
11909             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11910             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11911             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11912             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11913                                 Luxury      Midsize          Sedan   

       highway MPG  city mpg  Popularity   MSRP  
0               26        19        3916  46135  
1               28        19        3916  40650  
2               28        20        3916  36350  
3               28        18        3916  29450  
4               28        18        3916  34500  
...            ...       ...         ...    ...  
11909           23        16         204  46120  
11910           23        16         204  56670  
11911           23        16         204  50620  
11912           23        16         204  50920  
11913           26        17          61  28995  

[11914 rows x 16 columns]

## Answer to question number 2
There are 11914 records in the data set.

In [78]:
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [79]:
type(df)

pandas.core.frame.DataFrame

In [80]:
df [
    df['Make']== 'BMW'
]

Make       Model  Year             Engine Fuel Type  Engine HP  \
0      BMW  1 Series M  2011  premium unleaded (required)      335.0   
1      BMW    1 Series  2011  premium unleaded (required)      300.0   
2      BMW    1 Series  2011  premium unleaded (required)      300.0   
3      BMW    1 Series  2011  premium unleaded (required)      230.0   
4      BMW    1 Series  2011  premium unleaded (required)      230.0   
...    ...         ...   ...                          ...        ...   
11901  BMW          Z4  2016  premium unleaded (required)      240.0   
11902  BMW          Z4  2016  premium unleaded (required)      335.0   
11903  BMW          Z8  2001  premium unleaded (required)      394.0   
11904  BMW          Z8  2002  premium unleaded (required)      394.0   
11905  BMW          Z8  2003  premium unleaded (required)      394.0   

       Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0                   6.0            MANUAL  rear wheel drive              2.0   
1                   6.0            MANUAL  rear wheel drive              2.0   
2                   6.0            MANUAL  rear wheel drive              2.0   
3                   6.0            MANUAL  rear wheel drive              2.0   
4                   6.0            MANUAL  rear wheel drive              2.0   
...                 ...               ...               ...              ...   
11901               4.0            MANUAL  rear wheel drive              2.0   
11902               6.0  AUTOMATED_MANUAL  rear wheel drive              2.0   
11903               8.0            MANUAL  rear wheel drive              2.0   
11904               8.0            MANUAL  rear wheel drive              2.0   
11905               8.0            MANUAL  rear wheel drive              2.0   

                             Market Category Vehicle Size Vehicle Style  \
0      Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                         Luxury,Performance      Compact   Convertible   
2                    Luxury,High-Performance      Compact         Coupe   
3                         Luxury,Performance      Compact         Coupe   
4                                     Luxury      Compact   Convertible   
...                                      ...          ...           ...   
11901                     Luxury,Performance      Compact   Convertible   
11902                Luxury,High-Performance      Compact   Convertible   
11903         Exotic,Luxury,High-Performance      Compact   Convertible   
11904         Exotic,Luxury,High-Performance      Compact   Convertible   
11905         Exotic,Luxury,High-Performance      Compact   Convertible   

       highway MPG  city mpg  Popularity    MSRP  
0               26        19        3916   46135  
1               28        19        3916   40650  
2               28        20        3916   36350  
3               28        18        3916   29450  
4               28        18        3916   34500  
...            ...       ...         ...     ...  
11901           34        22        3916   49700  
11902           24        17        3916   66350  
11903           19        12        3916  128000  
11904           19        12        3916  130000  
11905           19        12        3916  131500  

[334 rows x 16 columns]

## Question number 3
Who are the most frequent car manufacturers (top-3) according to the dataset?

In [81]:
df.nunique() #gives number of unique values in each series/coloumn

Make                   48
Model                 915
Year                   28
Engine Fuel Type       10
Engine HP             356
Engine Cylinders        9
Transmission Type       5
Driven_Wheels           4
Number of Doors         3
Market Category        71
Vehicle Size            3
Vehicle Style          16
highway MPG            59
city mpg               69
Popularity             48
MSRP                 6049
dtype: int64

In [82]:
df['Make'].unique()

array(['BMW', 'Audi', 'FIAT', 'Mercedes-Benz', 'Chrysler', 'Nissan',
       'Volvo', 'Mazda', 'Mitsubishi', 'Ferrari', 'Alfa Romeo', 'Toyota',
       'McLaren', 'Maybach', 'Pontiac', 'Porsche', 'Saab', 'GMC',
       'Hyundai', 'Plymouth', 'Honda', 'Oldsmobile', 'Suzuki', 'Ford',
       'Cadillac', 'Kia', 'Bentley', 'Chevrolet', 'Dodge', 'Lamborghini',
       'Lincoln', 'Subaru', 'Volkswagen', 'Spyker', 'Buick', 'Acura',
       'Rolls-Royce', 'Maserati', 'Lexus', 'Aston Martin', 'Land Rover',
       'Lotus', 'Infiniti', 'Scion', 'Genesis', 'HUMMER', 'Tesla',
       'Bugatti'], dtype=object)

In [83]:
df['Make'].value_counts()  # this gives the frequencies of each unique value in the series 'Make'

Chevrolet        1123
Ford              881
Volkswagen        809
Toyota            746
Dodge             626
Nissan            558
GMC               515
Honda             449
Mazda             423
Cadillac          397
Mercedes-Benz     353
Suzuki            351
BMW               334
Infiniti          330
Audi              328
Hyundai           303
Volvo             281
Subaru            256
Acura             252
Kia               231
Mitsubishi        213
Lexus             202
Buick             196
Chrysler          187
Pontiac           186
Lincoln           164
Oldsmobile        150
Land Rover        143
Porsche           136
Saab              111
Aston Martin       93
Plymouth           82
Bentley            74
Ferrari            69
FIAT               62
Scion              60
Maserati           58
Lamborghini        52
Rolls-Royce        31
Lotus              29
Tesla              18
HUMMER             17
Maybach            16
McLaren             5
Alfa Romeo          5
Spyker    

### Answer to question number 3
The most frequent car manufacturers (top-3) according to the dataset are: Chevrolet, Ford and Volkswagen
____________

## Question number 4
Number of unique Audi car models

In [84]:
df[
    df['Make']== 'Audi'
]['Model'].nunique()

34

###  Answer to question number 4
Thus, number of unique Audi car models is 34
____

## Question number 5
Number of columns with missing values

In [85]:
df.isnull()

Make  Model   Year  Engine Fuel Type  Engine HP  Engine Cylinders  \
0      False  False  False             False      False             False   
1      False  False  False             False      False             False   
2      False  False  False             False      False             False   
3      False  False  False             False      False             False   
4      False  False  False             False      False             False   
...      ...    ...    ...               ...        ...               ...   
11909  False  False  False             False      False             False   
11910  False  False  False             False      False             False   
11911  False  False  False             False      False             False   
11912  False  False  False             False      False             False   
11913  False  False  False             False      False             False   

       Transmission Type  Driven_Wheels  Number of Doors  Market Category  \
0                  False          False            False            False   
1                  False          False            False            False   
2                  False          False            False            False   
3                  False          False            False            False   
4                  False          False            False            False   
...                  ...            ...              ...              ...   
11909              False          False            False            False   
11910              False          False            False            False   
11911              False          False            False            False   
11912              False          False            False            False   
11913              False          False            False            False   

       Vehicle Size  Vehicle Style  highway MPG  city mpg  Popularity   MSRP  
0             False          False        False     False       False  False  
1             False          False        False     False       False  False  
2             False          False        False     False       False  False  
3             False          False        False     False       False  False  
4             False          False        False     False       False  False  
...             ...            ...          ...       ...         ...    ...  
11909         False          False        False     False       False  False  
11910         False          False        False     False       False  False  
11911         False          False        False     False       False  False  
11912         False          False        False     False       False  False  
11913         False          False        False     False       False  False  

[11914 rows x 16 columns]

#### Let's check for missing values in all columns

In [86]:
df.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

#### Answer to Question number 5: we see above that 5 number of columns have missing values

____________________

## Question number 6
Does the median value change after filling missing values?

In [87]:
# let's first calculate median value of Engine Cylinders
df['Engine Cylinders'].median()  #here NaN are ignored for calculation by default

6.0

In [88]:
mode= df['Engine Cylinders'].mode()   #the most frequent number is mode by definition, NaN are ignored for calculation
mode

0    4.0
dtype: float64

In [89]:
df['Engine Cylinders'].isnull().sum()

30

In [90]:
# now lets fill the NaN in 'Engine Cylinders' column by the most frequent number
df['Engine Cylinders'].fillna(mode[0],inplace=True)
df['Engine Cylinders'].isnull().sum()

0

In [91]:
df['Engine Cylinders'].median() 

6.0

### Answer to question number 6
Median value doesn't change after filling missing values
____________

## Question number 7
- Select all the "Lotus" cars from the dataset.
- Select only columns "Engine HP", "Engine Cylinders".
- Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).
- Get the underlying NumPy array. Let's call it X
- Compute matrix-matrix multiplication between the transpose of $X$ and $X$. To get the transpose, use X.T. Let's call the result $XTX$.
- Invert XTX.
- Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
- Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
- What's the value of the first element of w?

In [95]:
df[df['Make']=='Lotus']  # selects all 'Lotus' cars from the dataset

Make      Model  Year                Engine Fuel Type  Engine HP  \
3912  Lotus      Elise  2009     premium unleaded (required)      189.0   
3913  Lotus      Elise  2009     premium unleaded (required)      218.0   
3914  Lotus      Elise  2009     premium unleaded (required)      189.0   
3915  Lotus      Elise  2010     premium unleaded (required)      189.0   
3916  Lotus      Elise  2010     premium unleaded (required)      218.0   
3917  Lotus      Elise  2011     premium unleaded (required)      189.0   
3918  Lotus      Elise  2011     premium unleaded (required)      217.0   
3919  Lotus      Elise  2011     premium unleaded (required)      217.0   
4216  Lotus     Esprit  2002     premium unleaded (required)      350.0   
4217  Lotus     Esprit  2003     premium unleaded (required)      350.0   
4218  Lotus     Esprit  2004     premium unleaded (required)      350.0   
4257  Lotus  Evora 400  2017     premium unleaded (required)      400.0   
4258  Lotus  Evora 400  2017     premium unleaded (required)      400.0   
4259  Lotus      Evora  2011     premium unleaded (required)      276.0   
4260  Lotus      Evora  2011     premium unleaded (required)      276.0   
4261  Lotus      Evora  2013     premium unleaded (required)      276.0   
4262  Lotus      Evora  2013     premium unleaded (required)      345.0   
4263  Lotus      Evora  2013     premium unleaded (required)      345.0   
4264  Lotus      Evora  2013     premium unleaded (required)      276.0   
4265  Lotus      Evora  2014     premium unleaded (required)      345.0   
4266  Lotus      Evora  2014     premium unleaded (required)      276.0   
4267  Lotus      Evora  2014     premium unleaded (required)      345.0   
4268  Lotus      Evora  2014     premium unleaded (required)      276.0   
4292  Lotus      Exige  2009  premium unleaded (recommended)      257.0   
4293  Lotus      Exige  2009  premium unleaded (recommended)      240.0   
4294  Lotus      Exige  2010  premium unleaded (recommended)      240.0   
4295  Lotus      Exige  2011  premium unleaded (recommended)      240.0   
4296  Lotus      Exige  2011  premium unleaded (recommended)      257.0   
4297  Lotus      Exige  2011  premium unleaded (recommended)      257.0   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
3912               4.0            MANUAL  rear wheel drive              2.0   
3913               4.0            MANUAL  rear wheel drive              2.0   
3914               4.0            MANUAL  rear wheel drive              2.0   
3915               4.0            MANUAL  rear wheel drive              2.0   
3916               4.0            MANUAL  rear wheel drive              2.0   
3917               4.0            MANUAL  rear wheel drive              2.0   
3918               4.0            MANUAL  rear wheel drive              2.0   
3919               4.0            MANUAL  rear wheel drive              2.0   
4216               8.0            MANUAL  rear wheel drive              2.0   
4217               8.0            MANUAL  rear wheel drive              2.0   
4218               8.0            MANUAL  rear wheel drive              2.0   
4257               6.0            MANUAL  rear wheel drive              2.0   
4258               6.0            MANUAL  rear wheel drive              2.0   
4259               6.0            MANUAL  rear wheel drive              2.0   
4260               6.0            MANUAL  rear wheel drive              2.0   
4261               6.0            MANUAL  rear wheel drive              2.0   
4262               6.0            MANUAL  rear wheel drive              2.0   
4263               6.0            MANUAL  rear wheel drive              2.0   
4264               6.0            MANUAL  rear wheel drive              2.0   
4265               6.0            MANUAL  rear wheel drive              2.0   
4266               6.0            MANUAL  rear wheel drive              2.0   
4267              

In [100]:
df1= df[df['Make']=='Lotus'][['Engine HP','Engine Cylinders']] #selects only columns "Engine HP", "Engine Cylinder
df1

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3914      189.0               4.0
3915      189.0               4.0
3916      218.0               4.0
3917      189.0               4.0
3918      217.0               4.0
3919      217.0               4.0
4216      350.0               8.0
4217      350.0               8.0
4218      350.0               8.0
4257      400.0               6.0
4258      400.0               6.0
4259      276.0               6.0
4260      276.0               6.0
4261      276.0               6.0
4262      345.0               6.0
4263      345.0               6.0
4264      276.0               6.0
4265      345.0               6.0
4266      276.0               6.0
4267      345.0               6.0
4268      276.0               6.0
4292      257.0               4.0
4293      240.0               4.0
4294      240.0               4.0
4295      240.0               4.0
4296      257.0               4.0
4297      257.0               4.0

In [105]:
df2=df1.drop_duplicates('Engine HP')
df2

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3918      217.0               4.0
4216      350.0               8.0
4257      400.0               6.0
4259      276.0               6.0
4262      345.0               6.0
4292      257.0               4.0
4293      240.0               4.0

In [112]:
x = df2.to_numpy() #dataframe is converted into an array
x

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

In [121]:
# transpose of x
xT=df2.T.to_numpy()
xT

array([[189., 218., 217., 350., 400., 276., 345., 257., 240.],
       [  4.,   4.,   4.,   8.,   6.,   6.,   6.,   4.,   4.]])

In [123]:
XTX =xT.dot(x)
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

In [124]:
np.linalg.det(XTX)  # just to check that the matrix is invertible

4556268.000000008

In [133]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

In [127]:
XTX.dot(np.linalg.inv(XTX))  #just to check that we get an identity matrix from this product

array([[ 1.00000000e+00, -1.37778677e-13],
       [-2.94902991e-17,  1.00000000e+00]])

In [132]:
# Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])
y

array([1100,  800,  750,  850, 1300, 1000, 1000, 1300,  800])

In [135]:
# Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
A = XTX_inv.dot(xT)
A

array([[-1.31950096e-03,  2.84443321e-04,  2.29134897e-04,
        -4.18763778e-03,  4.46417989e-03, -2.39406462e-03,
         1.42221660e-03,  2.44147184e-03,  1.50122864e-03],
       [ 8.60893170e-02,  7.36567735e-04,  3.67976598e-03,
         2.54588185e-01, -2.13748621e-01,  1.51207962e-01,
        -5.18727169e-02, -1.14048164e-01, -6.40137937e-02]])

In [136]:
w = A.dot(y)
w

array([  4.59494481, -63.56432501])

## Answer to question number 7:
The value of first element of w is 4.5949
________________